### ISSUE TITLE
Question about the intermediate-level features from the support image input into the multi-scale decoder

### ISSUE THREAD

**Issue by bowen099** [amirrezafateh/MSDNet#2]
您好，感谢你的开源项目。
请问在多尺度解码器中这里与query_feat拼接的量为supp_feat_cat_120， cat = torch.cat((query_feat,supp_feat_cat_120), dim =1)
而我发现在代码中supp_feat_cat_120的来源为： supp_feat_cat_120=self.convup2(supp_feat_mask_list[0])
而supp_feat_mask_list[0]得来源为：

![Image](https://github.com/user-attachments/assets/34d1976c-d4f5-45a6-9b0c-c8b3d2eb9c8d)
这是经过mask限定区域之后得中层特征，而不是原本得中间层次特征。

这与论文的图解是不是有所出入：

![Image](https://github.com/user-attachments/assets/bb6f95ab-7625-4df6-a44d-6c0ae1b023e4)

感谢您的指点

**Comment by amirrezafateh** [comment]
Hello, and thank you for your question!

I used translation because your question was not in English. If I misunderstood anything, please feel free to clarify.

Yes, the feature you pointed out is indeed constrained by the mask. Having masking processing is a common practice in similar architectures.

Your observation is valuable, and I appreciate the careful analysis! Let me know if you have any further questions.

**Comment by bowen099** [comment]
That is to say, the mid-level merged features of the support image input to the multi-scale decoder are indeed constrained by the mask (the part I underlined), but this is not explicitly indicated in the illustration in the paper. However, the high-level features of the support image input to the multi-scale decoder (i.e., the output of the last layer) are not combined with the mask, correct?

**Comment by amirrezafateh** [comment]
No, masking is a common practice, and in most cases, we should apply it to the feature maps of the support image. This applies not only to mid-level features but also to high-level features. There is a part in the code where high-level support features are used, and masking is explicitly applied in that operation.

Specifically, in the CMGM module, high-level support features are processed with a mask:

> corr_query_mask = self.CMGM(query_feat_high, supp_feat_high, s_y, fts_size)

Here, s_y is the mask, and inside CMGM, the masking operation is performed as follows:

> tmp_mask = (s_y[:, st, :, :] == 1).float().unsqueeze(1)
> tmp_mask = F.interpolate(tmp_mask, size=(fts_size[0], fts_size[1]), mode='bilinear', align_corners=True)
> tmp_supp_feat = supp_feat_high[:, st, ...] * tmp_mask

This ensures that the high-level support features are also constrained by the mask when they are used in CMGM.

**Comment by bowen099** [comment]
This line is not constrained by the mask, right?

![Image](https://github.com/user-attachments/assets/cc131066-fcee-4736-b8ae-b34b2c830fd3)

**Comment by bowen099** [comment]
![Image](https://github.com/user-attachments/assets/9e6263ab-519b-4c9f-a7ca-2a139854bc3c)
Thank you for your response. I just want to double-check—have the red lines in the past two days been constrained by the mask?

**Comment by bowen099** [comment]
![Image](https://github.com/user-attachments/assets/9da95444-70c4-499a-ba2c-8a9be582523f)
Thank you for your response. I just want to double-check—are these two red lines constrained by the mask?

**Comment by amirrezafateh** [comment]
Yes, just to make sure you're using the correct code, you should uncomment the two sections that are currently commented out for multi-scaling. I'll also add this point to the README.
